httpbp/client.go

```diff
@@ -0,0 +1,164 @@
+package httpbp
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// LowLevelClientMiddleware is suitable for transport-level functions.  Retries
+// and request munging should be done by wrapping the client.
+type LowLevelClientMiddleware func(name string, next http.RoundTripper) http.RoundTripper
+
+type maxConcurrency struct {
+	next           http.RoundTripper
+	activeRequests int64
+	maxConcurrency int64
+}
+
+func (m *maxConcurrency) RoundTrip(req *http.Request) (*http.Response, error) {
+	attemptedRequests := atomic.AddInt64(&m.activeRequests, 1)
+	defer atomic.AddInt64(&m.activeRequests, -1)
+	if m.maxConcurrency > 0 && attemptedRequests > m.maxConcurrency {
+		return nil, errors.New("hit concurrency limit on http round-tripper")
+	}
+	return m.next.RoundTrip(req)
+}
+
+// MaxConcurrency bounds the total number of requests in-flight, erroring if the
+// limit is exceeded.
+func MaxConcurrency(max int64) LowLevelClientMiddleware {
+	return func(name string, next http.RoundTripper) http.RoundTripper {
+		return &maxConcurrency{
+			next:           next,
+			maxConcurrency: max,
+		}
+	}
+}
+
+type circuitBreaker struct {
+	next    http.RoundTripper
+	breaker *gobreaker.CircuitBreaker
+}
+
+func (c *circuitBreaker) RoundTrip(req *http.Request) (*http.Response, error) {
+	rsp, err := c.breaker.Execute(func() (interface{}, error) {
+		r, e := c.next.RoundTrip(req)
+		if e != nil {
+			return nil, e
+		}
+		if r.StatusCode > 499 {
+			e = r.Body.Close()
+			if e != nil {
+				return nil, e
+			}
+			return nil, fmt.Errorf("received http %d", r.StatusCode)
```

Comment:
Converting 5xx status code into errors seems against how `RoundTripper` is defined: https://pkg.go.dev/net/http?tab=doc#RoundTripper

```
	// RoundTrip should not attempt to interpret the response. In
	// particular, RoundTrip must return err == nil if it obtained
	// a response, regardless of the response's HTTP status code.
	// A non-nil err should be reserved for failure to obtain a
	// response.
```

Reply:
Yeah, at this point, have we run into enough issues with trying to build at this layer that we'd prefer to just wrap a full http client, rather than try to mess with RoundTripper?

Reply:
If converting 5xx codes into errors is a hard requirement, then `RoundTripper` is probably the wrong solution for this.